src/js_mixed.hpp

```diff
@@ -0,0 +1,187 @@
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <common/type_deduction.hpp>
+#include <iostream>
+#include <map>
+#include <realm/mixed.hpp>
+#include <type_traits>
+
+namespace realm {
+namespace js {
+
+template <typename Context, typename Value>
+class MixedWrapper {
+   public:
+    virtual Mixed wrap(Context, Value const &) = 0;
+    virtual Value unwrap(Context, Mixed) = 0;
+};
+
+template <typename Context, typename Value, typename Utils>
+class HandleString : public MixedWrapper<Context, Value> {
+   private:
+    // we need this <cache> to keep the value life long enough to get into the
+    // DB, we do this because the realm::Mixed type is just a reference
+    // container.
+    std::string cache;
+
+   public:
+    Mixed wrap(Context context, Value const &value) {
+        cache = Utils::to_string(context, value);
+        return Mixed(cache);
+    }
+
+    Value unwrap(Context context, Mixed mixed) {
+        return Utils::from_string(context,
+                                  mixed.template get<StringData>().data());
+    }
+};
+
+template <typename Context, typename Value, typename Utils>
+class HandleBoolean : public MixedWrapper<Context, Value> {
+    Mixed wrap(Context context, Value const &value) {
+        return Mixed(Utils::to_boolean(context, value));
+    }
+
+    Value unwrap(Context context, Mixed mixed) {
+        return Utils::from_boolean(context, mixed.get<bool>());
+    }
+};
+
+template <typename Context, typename Value, typename Utils,
+          typename RealmNumberType>
+class HandleNumber : public MixedWrapper<Context, Value> {
+    Mixed wrap(Context context, Value const &value) {
+        return Mixed(Utils::to_number(context, value));
+    }
+
+    Value unwrap(Context context, Mixed mixed) {
+        return Utils::from_number(context, mixed.get<RealmNumberType>());
+    }
+};
+
+template <typename Context, typename Value, typename Utils>
+class HandleDecimal128 : public MixedWrapper<Context, Value> {
+    Mixed wrap(Context context, Value const &value) {
+        return Mixed(Utils::to_decimal128(context, value));
+    }
+
+    Value unwrap(Context context, Mixed mixed) {
+        return Utils::from_decimal128(context, mixed.get<Decimal>());
+    }
+};
+
+template <typename Context, typename Value, typename Utils>
+class HandleObjectID : public MixedWrapper<Context, Value> {
+    Mixed wrap(Context context, Value const &value) {
+        return Mixed(Utils::to_object_id(context, value));
+    }
+
+    Value unwrap(Context context, Mixed mixed) {
+        return Utils::from_object_id(context, mixed.get<ObjectId>());
+    }
+};
+
+template <typename Context, typename Value, typename Utils>
+class HandleBinary : public MixedWrapper<Context, Value> {
+    Mixed wrap(Context context, Value const &value) {
+        auto owned_binary_data = Utils::to_binary(context, value);
```

Comment:
`BinaryData` and `StringData` are very similar in Core - they don't hold a reference to their contents. So it's surprising that you need the cache for the string case, but not in the binary case.

Reply:
Good catch, thanks for the heads up 👍